### ISSUE TITLE
Function to register new textobjects

### ISSUE THREAD

**Issue by GCBallesteros** [echasnovski/mini.nvim#358]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)

### Module(s)

mini.ai

### Description

Add a function to easily register new textobjects like one can do in [vim-textobj-user](https://github.com/kana/vim-textobj-user).

I'm writing a plugin that works on text objects for which I've defined a mini.ai compatible function. This makes it possible for users to use this textobject outside the confines of the plugin. The way I'm doing it right now is as follows:

```lua
-- the config for mini.ai has already happened before
local nn = require "notebook-navigator" -- <- my plugin
nn.setup({})
_G.MiniAi.config.custom_textobjects.h = nn.notebook_navigator_textobject
```

and it works just fine but I think it would be even nicer to be able to do something like:

```lua
local ai = require "mini.ai"
local nn = require "notebook-navigator"
ai.register_textobject("h", nn.notebook_navigator_textobject)
```

I think that this kind of functionality would also promote the development of a mini ecosystem of user defined textobjects [like you get with vim-textobj-user](https://github.com/kana/vim-textobj-user/wiki#general-purpose-text-objects).

I wouldn't mind working on a PR if this feature is deemed appropriate for inclusion into mini.ai

--------

Edit:
TL;DR The way of doing things described above by changing `_G.MiniAi` will cause you grief. Don't do it and instead follow  the advice from @echasnovski  below.

Modifying `_G.MiniAi.config` in your configs may uncover non deterministic behaviour from your package manager if you are not extra careful. For it to work you would have had to run the setup for "mini.ai.setup" or otherwise `_G.MiniAi` simply won't exist that much is obvious. If you use lazy.nvim it would seem like putting "echasnovski/mini.ai" on the dependencies would suffice but that's not quite true by itself. If for example your configuration for 'mini.ai' uses  'mini.nvim' (and then calls `require "mini.ai"`) as for example in:

```lua
{
  "echasnovski/mini.nvim",
   config = function()
     local ai = require "mini.ai"
     ai.setup({})
  end
}
```

 but your dependencies on lazy, as stated above are for "mini.ai", as in:
 ```lua
{
  "GCBallesteros/NotebookNavigator.nvim",
  dependencies = {"echasnovski/mini.ai"},  -- <--- DANGER!
  config = function()
    local nn = require "notebook-navigator"
    nn.setup({})
    _G.MiniAi.config.custom_textobjects.h = nn.notebook_navigator_textobject
  end
}
```
then lazy won't respect the dependency (because it's not really there) and your config will fail on startup randomly.

**Comment by echasnovski** [comment]
Thanks for the idea!

At the moment I don't think this introduces *that* different approach from the current one. How I currently assume things should work:
- User installs 'mini.ai' and 'notebook-navigator'.
- Has the following setup for 'mini.ai':
```lua
require('mini.ai').setup({
  custom_textobjects = {
    h = require('notebook-naviagtor').notebook_navigator_textobject
  }
})
```

------

On a side note, I'd suggest using "textobject specification" or just "specification" instead of plain "textobject". Maybe something like `require('notebook-naviagtor').miniai_spec`?

**Comment by GCBallesteros** [comment]
Thanks for responding so quickly.

I thought about the approach you are proposing above but chose not to go that way because it would mean that now the mini.ai configuration depends on a plugin from some random person. This is particularly noticeable when you are specifying   package dependencies upfront (as you do in packer or lazy). Now for each textobject specification I want to bring from outside, my mini.ai config adds one dependency and to me at least that feels a bit off.

Let me know if you change your mind on the topic and thanks again for all the amazing plugins.

P.S. I'm taking your advice above on naming conventions

**Comment by echasnovski** [comment]
> I thought about the approach you are proposing above but chose not to go that way because it would mean that now the mini.ai configuration depends on a plugin from some random person. This is particularly noticeable when you are specifying package dependencies upfront (as you do in packer or lazy). Now for each textobject specification I want to bring from outside, my mini.ai config adds one dependency and to me at least that feels a bit off.

The alternative would be to specify 'mini.ai' as dependency for *every* plugin implementing textobject specification for 'mini.ai', which to me feels a bit off.

Having "external" plugins as dependency in one place (for 'mini.ai') seems more straightforward as it enables explicit configuration of `a`/`i` textobjects one place. Similar to relation of 'nvim-cmp' and its sources.

**Comment by GCBallesteros** [comment]
> The alternative would be to specify 'mini.ai' as dependency for every plugin implementing textobject specification for 'mini.ai', which to me feels a bit off.

And to me feels more natural :D. I'm coming from the perspective of using 'mini.ai' as a sort of library while your point of view seems (to me) to be that 'mini.ai' is more an application that takes on dependencies. I'm afraid I don't have any particularly strong argument here so I will let it be, and then of course, your house your rules.

Thanks for entertaining me and please do let me know if you'd like somebody to work on such a feature.

**Comment by echasnovski** [comment]
After pondering over it for a while, I think my initial suggestion on how to organize external textobject stands. Basically, put them as dependency of 'mini.ai'/'mini.nvim' and explicitly use them inside `require('mini.ai').setup({ custom_textobjects = { ... } })`.

Nevertheless, thanks for your efforts to create custom textobject specification!
